src/gui/fspathedit.cpp

```diff
@@ -0,0 +1,361 @@
+        break;
+    case FileSystemPathEdit::Mode::DirectoryOpen:
+    case FileSystemPathEdit::Mode::DirectorySave:
+        selectedPath = QFileDialog::getExistingDirectory(q, dialogCaptionOrDefault(),
+                                directory, QFileDialog::DontResolveSymlinks | QFileDialog::ShowDirsOnly);
+        break;
+    default:
+        throw std::logic_error("Unknown FileSystemPathEdit mode");
+    }
+    if (!selectedPath.isEmpty())
+        q->setEditWidgetText(selectedPath);
+}
+
+QString FileSystemPathEdit::FileSystemPathEditPrivate::dialogCaptionOrDefault() const
+{
+    if (!m_dialogCaption.isEmpty())
+        return m_dialogCaption;
+
+    switch (m_mode) {
+    case FileSystemPathEdit::Mode::FileOpen:
+    case FileSystemPathEdit::Mode::FileSave:
+        return defaultDialogCaptionForFile.tr();
+    case FileSystemPathEdit::Mode::DirectoryOpen:
+    case FileSystemPathEdit::Mode::DirectorySave:
+        return defaultDialogCaptionForDirectory.tr();
+    default:
+        throw std::logic_error("Unexpected value of type FileSystemPathEdit::Mode");
+    }
+}
+
+void FileSystemPathEdit::FileSystemPathEditPrivate::modeChanged()
+{
+    QStyle::StandardPixmap pixmap = QStyle::SP_DialogOpenButton;
+    bool showDirsOnly = false;
+    switch (m_mode) {
+    case FileSystemPathEdit::Mode::FileOpen:
+    case FileSystemPathEdit::Mode::FileSave:
+        pixmap = QStyle::SP_DialogOpenButton;
+        showDirsOnly = false;
+        break;
+    case FileSystemPathEdit::Mode::DirectoryOpen:
+    case FileSystemPathEdit::Mode::DirectorySave:
+        pixmap = QStyle::SP_DirOpenIcon;
+        showDirsOnly = true;
+        break;
+    }
+    m_browseAction->setIcon(QApplication::style()->standardIcon(pixmap));
+    m_editor->completeDirectoriesOnly(showDirsOnly);
+}
+
+FileSystemPathEdit::FileSystemPathEdit(Private::FileEditorWithCompletion *editor, QWidget *parent)
+    : QWidget(parent)
+    , d_ptr(new FileSystemPathEditPrivate(this, editor))
+{
+    Q_D(FileSystemPathEdit);
+    editor->widget()->setParent(this);
+
+    QHBoxLayout *layout = new QHBoxLayout(this);
+    layout->addWidget(editor->widget());
+    layout->addWidget(d->m_browseBtn);
+
+    connect(d->m_browseAction, &QAction::triggered, [this]() {this->d_func()->browseActionTriggered();});
+}
+
+FileSystemPathEdit::~FileSystemPathEdit() = default;
+
+QString FileSystemPathEdit::selectedPath() const
+{
+    return Utils::Fs::fromNativePath(editWidgetText());
+}
+
+void FileSystemPathEdit::setSelectedPath(const QString &val)
+{
+    Q_D(FileSystemPathEdit);
+    setEditWidgetText(Utils::Fs::toNativePath(val));
+    d->m_editor->widget()->setToolTip(val);
+}
+
+QString FileSystemPathEdit::fileNameFilter() const
+{
+    Q_D(const FileSystemPathEdit);
+    return d->m_fileNameFilter;
+}
+
+void FileSystemPathEdit::setFileNameFilter(const QString &val)
+{
+    Q_D(FileSystemPathEdit);
+    d->m_fileNameFilter = val;
+
+#if 0
```

Comment:
Yes, `#if 0` is perfect for commenting out a block of code which itself contains comments inside.